utils/diffing_with_keys.ml

```diff
@@ -57,19 +57,32 @@ module Swap = Map.Make(struct
   end)
 module Move = Misc.Stdlib.String.Map
 
-type ('a,'state) partial_edge =
-  | Left of int * 'state * 'a
-  | Right of int * 'state * 'a
-  | Both of 'state * 'a * 'a
+type ('l, 'r, 'state) partial_edge =
+  | Left of int * 'state * 'l
+  | Right of int * 'state * 'r
+  | Both of 'state * 'l * 'r
 
-let edge key state x y =
-  let kx, ky = key (data x), key (data y) in
+type ('l, 'r) keys = ('l -> string) * ('r -> string)
+
+let swap_edge (type l r state)
+    (key : (l, r) keys)
+    (state : state)
+    (x : l with_pos)
+    (y : r with_pos)
+  : Swap.key
+    * (l with_pos * r with_pos, l with_pos * r with_pos, state) partial_edge
```

Comment:
Note: I am not sure whether the type of swaps should be `(l * r, l * r) partial_edge` or rather `(l * r, r * l) partial_erge`.